# CA1 - Data Preparation
**Student name:** Gleiciane Rocha dos Santos
**Student ID:** 2026248  
**Module:** Data Preparation  
**Programme:** HDIP Data Analytics  
**Dataset:** student_data.csv




## Introduction

This notebook looks at the dataset `student_data.csv`, which contains information about students and their outcomes, the aim of this analysis is to understand the dataset, improve its quality, prepare it for analysis, identify patterns in the data, and use PCA to reduce dimensionality while keeping most of the information. 
It also discusses the curse of dimensionality and checks whether the reduced dataset is still useful in practice.

## Dataset Characterisation

In [17]:
import pandas as pd
import numpy as np

In [20]:
df = pd.read_csv("student_data.csv")

The dataset was loaded into Python using pandas

In [21]:
df.shape

(4424, 37)

The dataset has 4424 rows and 37 columns. This gives a general idea of the size of the data used in this analysis

In [23]:
df.columns

Index(['Marital status', 'Application mode', 'Application order', 'Course',
       'Daytime/evening attendance', 'Previous qualification',
       'Previous qualification (grade)', 'Nacionality',
       'Mother's qualification', 'Father's qualification',
       'Mother's occupation', 'Father's occupation', 'Admission grade',
       'Displaced', 'Educational special needs', 'Debtor',
       'Tuition fees up to date', 'Gender', 'Scholarship holder',
       'Age at enrollment', 'International',
       'Curricular units 1st sem (credited)',
       'Curricular units 1st sem (enrolled)',
       'Curricular units 1st sem (evaluations)',
       'Curricular units 1st sem (approved)',
       'Curricular units 1st sem (grade)',
       'Curricular units 1st sem (without evaluations)',
       'Curricular units 2nd sem (credited)',
       'Curricular units 2nd sem (enrolled)',
       'Curricular units 2nd sem (evaluations)',
       'Curricular units 2nd sem (approved)',
       'Curricular units 2nd s

The columns names were checked to understand wich varibles are included in the dataset and confirm the presence of the target variable 

In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4424 entries, 0 to 4423
Data columns (total 37 columns):
 #   Column                                          Non-Null Count  Dtype  
---  ------                                          --------------  -----  
 0   Marital status                                  4424 non-null   object 
 1   Application mode                                4424 non-null   int64  
 2   Application order                               4424 non-null   int64  
 3   Course                                          4423 non-null   float64
 4   Daytime/evening attendance                      4424 non-null   int64  
 5   Previous qualification                          4424 non-null   int64  
 6   Previous qualification (grade)                  4424 non-null   float64
 7   Nacionality                                     4424 non-null   int64  
 8   Mother's qualification                          4424 non-null   int64  
 9   Father's qualification                   

Info output helped to check the structure, especially variables types and any possible missing values

In [25]:
df.isna().sum()

Marital status                                    0
Application mode                                  0
Application order                                 0
Course                                            1
Daytime/evening attendance                        0
Previous qualification                            0
Previous qualification (grade)                    0
Nacionality                                       0
Mother's qualification                            0
Father's qualification                            1
Mother's occupation                               0
Father's occupation                               0
Admission grade                                   1
Displaced                                         0
Educational special needs                         0
Debtor                                            1
Tuition fees up to date                           1
Gender                                            0
Scholarship holder                                0
Age at enrol

Missing values were checked to see any variable had imcomplete data

In [26]:
df["Target"].value_counts(dropna=False)

Target
Graduate    2207
Dropout     1419
Enrolled     794
Grad           1
graduate       1
Drop           1
dropout        1
Name: count, dtype: int64

The variable `Target` shows the student outcome groups used in this analysis it includes the categories `Graduate`, `Dropout`, and `Enrolled`

In general, the characterization shows that the dataset is suitable for analysis, but it also presents some problems that require attention, such as missing values ​​and variables that need further verification. Therefore, a data preparation step is necessary before proceeding to the next phase of the analysis

## *Data Preparation*

In this step, the dataset is prepared for analyses by cheking the issues found ealier, looking for inconsistent values, variables types and cleaning some columns, so the data is more ready for the next stages of the analysis

In [29]:
df["Displaced"].value_counts(dropna=False)

Displaced
1    2425
0    1998
?       1
Name: count, dtype: int64

The variable 'displaced'is a binary variable so it's expected to contain only valid values such as '0' and '1' However, one inconsistent value was found, which means the column needs to be cleaned before the next steps

In [30]:
df["Displaced"] = df["Displaced"].replace("?", np.nan)

Replace the value '?' with 'NaN', for Python interprets it as a missing value

In [31]:
df["Displaced"] = pd.to_numeric(df["Displaced"])

Convert columns to number because being a binary variable, makes more sense as a numeric variable 

In [32]:
df["Displaced"].dtype

dtype('float64')

In [34]:
df["Displaced"].isna().sum()

np.int64(1)

The inconsistent value in `Displaced` was treated as a missing value, and the column was converted to numeric format

In [35]:
df["Tuition fees up to date"].value_counts(dropna=False)

Tuition fees up to date
1      3895
0       527
NaN       1
?         1
Name: count, dtype: int64

`Tuition fees up to date` is a binary variable, so it should only have `0` and `1`one inconsistent value was found, and the column also has a missing value, so it needs cleaning

In [36]:
df["Tuition fees up to date"] = df["Tuition fees up to date"].replace("?", np.nan)

In [37]:
df["Tuition fees up to date"] = pd.to_numeric(df["Tuition fees up to date"])

In [38]:
df["Tuition fees up to date"].dtype

dtype('float64')

In [40]:
df["Tuition fees up to date"].isna().sum()

np.int64(2)

The inconsistent value in `Tuition fees up to date` was treated as a missing value, and the column was converted to numeric format

In [41]:
df["Age at enrollment"].value_counts(dropna=False)

Age at enrollment
18         1036
19          911
20          599
21          321
22          174
24          131
23          108
26           94
25           93
27           91
28           83
29           66
32           61
34           60
31           55
35           49
30           49
33           45
37           42
39           38
36           35
41           31
38           29
43           25
40           23
45           22
44           21
47           18
50           16
42           13
49           13
46           12
48           11
54            7
53            7
51            7
55            5
17            5
52            4
58            3
59            3
57            2
60            2
UnKnown       1
61            1
70            1
62            1
Name: count, dtype: int64

`Age at enrollment` should only contain age values, but one inconsistent value (`UnKnown`) was found this needs to be cleaned before the next steps

In [42]:
df["Age at enrollment"] = df["Age at enrollment"].replace("UnKnown", np.nan)

In [43]:
df["Age at enrollment"] = pd.to_numeric(df["Age at enrollment"])

In [44]:
df["Age at enrollment"].dtype

dtype('float64')

In [45]:
df["Age at enrollment"].isna().sum()

np.int64(1)

The invalid value in `Age at enrollment` was changed to a missing value, and the column was converted to numeric format

In [46]:
df["Curricular units 1st sem (credited)"].value_counts(dropna=False)

Curricular units 1st sem (credited)
0     3847
2       94
1       85
3       69
6       50
4       47
5       41
7       41
8       31
9       27
11      17
10      15
14      15
13      13
12      12
15       5
18       4
17       3
16       3
19       2
20       2
Na       1
Name: count, dtype: int64

`Curricular units 1st sem (credited)` should only contain numeric values, but one inconsistent value (`Na`) was found this needs to be cleaned before the next steps

In [47]:
df["Curricular units 1st sem (credited)"] = df["Curricular units 1st sem (credited)"].replace("Na", np.nan)

In [48]:
df["Curricular units 1st sem (credited)"] = pd.to_numeric(df["Curricular units 1st sem (credited)"])

In [49]:
df["Curricular units 1st sem (credited)"].dtype

dtype('float64')

In [50]:
df["Curricular units 1st sem (credited)"].isna().sum()

np.int64(1)

The invalid value in `Curricular units 1st sem (credited)` was changed to a missing value and the column was converted to numeric format

In [51]:
df["Target"].value_counts(dropna=False)

Target
Graduate    2207
Dropout     1419
Enrolled     794
Grad           1
graduate       1
Drop           1
dropout        1
Name: count, dtype: int64

`Target` is a categorical variable, but some outcome labels are written in different ways because of this, the column needs to be cleaned before the next steps

In [52]:
df["Target"] = df["Target"].replace("graduate", "Graduate")
df["Target"] = df["Target"].replace("Grad", "Graduate")

In [53]:
df["Target"] = df["Target"].replace("dropout", "Dropout")
df["Target"] = df["Target"].replace("Drop", "Dropout")

In [54]:
df["Target"].value_counts(dropna=False)

Target
Graduate    2209
Dropout     1421
Enrolled     794
Name: count, dtype: int64

The labels in `Target` have been cleaned and standardized

In [55]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4424 entries, 0 to 4423
Data columns (total 37 columns):
 #   Column                                          Non-Null Count  Dtype  
---  ------                                          --------------  -----  
 0   Marital status                                  4424 non-null   object 
 1   Application mode                                4424 non-null   int64  
 2   Application order                               4424 non-null   int64  
 3   Course                                          4423 non-null   float64
 4   Daytime/evening attendance                      4424 non-null   int64  
 5   Previous qualification                          4424 non-null   int64  
 6   Previous qualification (grade)                  4424 non-null   float64
 7   Nacionality                                     4424 non-null   int64  
 8   Mother's qualification                          4424 non-null   int64  
 9   Father's qualification                   

After cleaning the main issues, the dataset was checked again using `info()`some columns were successfully converted to numeric format, so the dataset is now more organised for the next steps